# Advanced Partial Derivatives


Let's say we want to determine the volume of a cylinder, using a multivariate function: it would be $\pi$ times the squared radius times the length of the cylinder $v = \pi r^2 l$.

The partial derivative of the volume with respect to the length of the cylinder would be: $\frac{\partial v}{\partial l}\frac{d}{dl}(l)$, and because $l$, according to the power rule is equal to 1, we can say that the partial derivative of the volume with respect to the length of the cylinder is equal to $\pi r^2$.

In this case, the variable $r^2$ is being treated as a constant because $r^2$ is not being differentiated.

A change in $l$ corresponds to a change in volume by $\pi r^2$.

In [2]:
import sympy as sp
import math
import torch

In [4]:
def cylinder_volume(my_r, my_l):
    return math.pi * my_r**2 * my_l

Let's assume the radius is 3 meters and the length is 5 meters:

In [5]:
r = torch.tensor(3.).requires_grad_()
l = torch.tensor(5.).requires_grad_()

In [6]:
v = cylinder_volume(r, l)
v

tensor(141.3717, grad_fn=<MulBackward0>)

In [7]:
v.backward()

In [8]:
l.grad

tensor(28.2743)

We can demonstrate that $\frac{\partial v}{\partial l} = \pi r^2$:

In [9]:
math.pi * 3**2

28.274333882308138

This means that with $r = 3$, a change in the length of the cylinder will change the volume by $\approx 28.27 m^3$ for every meter of change in length.

In [10]:
cylinder_volume(3, 6)

169.64600329384882

In [11]:
cylinder_volume(3, 6) - cylinder_volume(3, 5)

28.274333882308127

In [17]:
cylinder_volume(3, 7) - cylinder_volume(3, 6)

28.274333882308156

Every increase of $l$ by $n$ units will increase the volume by $\pi r^2$ times $n$:

In [18]:
cylinder_volume(3, 8) - cylinder_volume(3, 6)

56.54866776461628

In [19]:
cylinder_volume(3, 9) - cylinder_volume(3, 6)

84.82300164692441

Let's also calculate the partial derivative of the volume with respect to the radius of the cylinder:

$$\frac{\partial v}{\partial r} = \pi\frac{d}{dr}(r^2)l = \pi(2r)l = 2\pi rl$$

What this means is that a change in the radius of the cylinder will change the volume by $2\pi rl$, but only if we work with an infinitesimal change in the radius.

In [24]:
r.grad

tensor(94.2478)

In [25]:
2 * math.pi * 3 * 5

94.24777960769379

Since $r$ is included in the partial derivative, adjusting it affects the scale of its impact on $v$

Typically, only an infinitesimally small $\Delta r$ would be applied. The smaller the $\Delta r$, the closer to the true $\frac{\partial v}{\partial r}$ we get. The larger the $\Delta r$, the more inaccurate the approximation becomes.

E.g., at $\Delta r = 1 \times 10^{-6}$:

In [29]:
delta_r = 1e-6
(cylinder_volume(3 + delta_r, 5) - cylinder_volume(3, 5)) / delta_r

94.24779531741478

Next up:

Find all the partial derivatives of the following functions:

1. $z = y^3 + 5xy$
2. The surface area of a cylinder is described by $a = 2\pi r^2 + 2\pi rh$.
3. The volume of a square prism with a cube cut out of its center is described by $v = x^2y - z^3$

In [34]:
# 1. partial derivatives of z = y^3 + 5xy
x, y, z = sp.symbols('x y z')
z = y**3 + 5*x*y
dz_dx = sp.diff(z, x)
dz_dy = sp.diff(z, y)
dz_dx

5*y

In [35]:
dz_dy

5*x + 3*y**2

In [36]:
# 2. partial derivatives of a = 2pi*r^2 + 2pi*rh
r, h, a = sp.symbols('r h a')
a = 2*sp.pi*r**2 + 2*sp.pi*r*h
da_dr = sp.diff(a, r)
da_dh = sp.diff(a, h)
da_dr

2*pi*h + 4*pi*r

In [37]:
da_dh

2*pi*r

In [38]:
# 3. partial derivatives of v = x^2y - z^3
x, y, z = sp.symbols('x y z')
v = x**2*y - z**3
dv_dx = sp.diff(v, x)
dv_dy = sp.diff(v, y)
dv_dz = sp.diff(v, z)
dv_dx

2*x*y

In [39]:
dv_dy

x**2

In [40]:
dv_dz

-3*z**2